### PR TITLE
Drop redundant dialogs during the traverse

### PR DESCRIPTION
**Pull request by mkondratek**

### CODE REVIEW

File: CHANGE-NOTES.md [hunk 1]
```diff
@@ -1,5 +1,8 @@
 # Changelog
 
+## v4.0.0
```
Comment (PawelLipski) [inline CHANGE-NOTES.md:3]: Nit: let's keep it v3.8.0 (or even v3.7.1, no new feature theoretically 🤓 ) for now

**Comment by PawelLipski** [comment]
Just tested now... there're a few visual glitches and ofc lack of integration with Yes+Quit feature... but generally :benissimo:

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java [hunk 2]
```diff
@@ -24,16 +25,20 @@
 
 public final class GitPushDialog extends VcsPushDialog {
   private final boolean isForcePushRequired;
+  @Nullable
+  private final JComponent traverseInfoComponent;
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java:29]: Nit:
```suggestion
  private final @Nullable JComponent traverseInfoComponent;
```

File: CHANGE-NOTES.md [hunk 3]
```diff
@@ -1,6 +1,7 @@
 # Changelog
 
-## v3.7.2
+## v3.8.0
```
Comment (PawelLipski) [inline CHANGE-NOTES.md:3]: Okay... I moved the issue & PR to v3.8.0 milestone

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java [hunk 4]
```diff
@@ -110,27 +130,67 @@ public void onSuccess() {
...
+      return isForcePushRequired ? "Force _Push" : "_Push";
+    }
   }
 
   private class PushSwingAction extends AbstractAction {
 
+    private final Runnable pushRunnable;
+
+    @UIEffect
+    PushSwingAction(boolean isAndQuit) {
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java:165]: Maybe `hasAndQuit`?

Comment (PawelLipski) [reply frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java:165]: Actually... maybe create 2 separate classes? `PushSwingAction` and `PushAndQuitSwingAction`

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java [hunk 5]
```diff
@@ -71,7 +85,13 @@ protected void doOKAction() {
   @Override
   @UIEffect
   protected Action[] createActions() {
-    return new Action[]{pushAction, getCancelAction(), getHelpAction()};
+    Action cancelAction = getCancelAction();
+    if (traverseInfoComponent != null) {
+      cancelAction.putValue(Action.NAME, "_Quit Traverse");
+      return new Action[]{skipAction, pushAction, pushActionAndQuit, cancelAction, getHelpAction()};
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java:91]: Hmmm

image

Is this button ordering consistent with the other traverse dialogs? 🤔

Comment (PawelLipski) [reply frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java:91]: `TraverseStepConfirmationDialog`:

```java
@Untainted String[] options = {yesText, yesAndQuitText, noText, quitText};
```

Comment (PawelLipski) [reply frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java:91]: Also, let's make the naming consistent (in particular, `Quit Traverse` is better than `Quit the interactive Traverse`)

File: frontend/resourcebundles/src/main/resources/GitMacheteBundle.properties [hunk 6]
```diff
@@ -545,26 +545,23 @@ action.GitMachete.BaseTraverseAction.dialog.pull-approval.text.HTML=\
     <p>Do you want to <b>pull</b> the remote changes?</p>\
   </html>
 
-action.GitMachete.BaseTraverseAction.dialog.push-approval.title=Git Machete Traverse - Push
```
[the remark this answers is not in the document]
Comment (mkondratek) [reply frontend/resourcebundles/src/main/resources/GitMacheteBundle.properties:548]: It's from IntelliJ applied on purpose for consistency. The original push dialog can be used on multiple repositories at once. I'd rather keep it to stick with the original dialog.

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java (hunk 5 above)
Comment (mkondratek) [reply frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java]: image

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java [hunk 7]
```diff
@@ -110,27 +129,80 @@ public void onSuccess() {
...
+    if (hasAndQuit) {
+      return (isForcePushRequired ? "Force Push" : "Push") + " _and Quit";
+    } else {
+      return isForcePushRequired ? "Force _Push" : "_Push";
+    }
   }
 
   private class PushSwingAction extends AbstractAction {
 
+    private final Runnable pushRunnable;
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java:161]: No need to extract `pushRunnable`, just call `push()` in `actionPerformed`

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java [hunk 8]
```diff
@@ -110,27 +129,80 @@ public void onSuccess() {
...
     @UIEffect
     public void actionPerformed(ActionEvent e) {
-      push();
+      pushRunnable.run();
+    }
+  }
+
+  private class PushAndQuitSwingAction extends AbstractAction {
+
+    private final Runnable pushRunnable;
```
Comment (PawelLipski) [inline frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java:179]: Ditto, field redundant

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java [hunk 9]
```diff
@@ -71,7 +84,13 @@ protected void doOKAction() {
   @Override
   @UIEffect
   protected Action[] createActions() {
-    return new Action[]{pushAction, getCancelAction(), getHelpAction()};
+    Action cancelAction = getCancelAction();
+    if (traverseInfoComponent != null) {
+      cancelAction.putValue(Action.NAME, getNonHtmlString("action.GitMachete.BaseTraverseAction.dialog.quit"));
+      return new Action[]{cancelAction, skipAction, pushAction, pushActionAndQuit, getHelpAction()};
+    } else {
+      return new Action[]{pushAction, cancelAction, getHelpAction()};
```
[the remark this answers is not in the document]
Comment (mkondratek) [reply frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java:92]: yes

[the remark this answers is not in the document]
Comment (mkondratek) [reply frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java:92]: the first case is traverse push - it has the git machete traverse message and context

[the remark this answers is not in the document]
Comment (mkondratek) [reply frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java:92]: the second is more like a regular push - there is no git machete description and context there actually. the only thing is that is automatically decides if force push is required

[the remark this answers is not in the document]
Comment (PawelLipski) [reply frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java:92]: Huh okay... still, somewhat surprising that we (or Intellij as a whole?) has two separate conventions for button order. Big Endian vs Little Endian :trollface:

**Review by PawelLipski — APPROVED** [review]

File: frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (mkondratek) [reply frontend/actions/src/main/java/com/virtuslab/gitmachete/frontend/actions/dialogs/GitPushDialog.java:92]: I think that 
```java
  @Override
  protected Action getOKAction() {
    return pushAction;
  }
```

defines which action should by rightmost eventually

**Pull request merged**